### ISSUE TITLE
Problem with GSM8K results

### ISSUE THREAD

**Issue by WJMacro** [sail-sg/sdft#15]
Hello, thank you for open-sourcing your code. While reproducing your work, I found that the experimental results on the GSM8K dataset is strange.

According to your paper, fine-tuning the GSM8K dataset using the original data achieved an accuracy of 31.9%. In contrast, fine-tuning with the paraphrased data reached 34.4%. While this appears to be an exciting result, the performance achieved by directly fine-tuning with the GSM8K data seems **unusually low**. 

Upon closely reviewing the experimental setup, I noticed that you used LoRA with a rank of 8, applied only to the query and value matrices. This experimental setup limits the model's learning capacity. Limited learning capacity might weaken the model's performance on the original dataset, making the fine-tuning baseline appear less competitive. 

To verify this hypothesis, I conducted experiments using your code. In my experiments, I made the following modifications to the training scripts:
```
    --lora_target all \
    --lora_rank 64 \
```

My experimental results are as follows:

sft result:
```
Fine-tuning using sft

Evaluation on gsm8k:
Accuracy for math: 561 / 1319 = 42.53%

Evaluation on multiarith:
Accuracy for math: 152 / 180 = 84.44%
```

sdft result:
```
Fine-tuning using sdft

Evaluation on gsm8k:
Accuracy for math: 456 / 1319 = 34.57%

Evaluation on multiarith:
Accuracy for math: 140 / 180 = 77.78%
```

According to the result, when the model has sufficient tunable parameters, the training performance with the original data (Vanilla-FT) improved significantly (from **31.9** to **42.5**). Conversely, for the paraphrased data (SDFT), increasing the number of trainable parameters did not improve performance (from **34.4** to **34.5**). Overall, SDFT underperforms Vanilla-FT when enlarging the trainable parameter. 
Additionally, since using a LoRA rank of 64 achieved a GSM8K accuracy of 42.5% with the original data, the accuracy of **34.87%** reported for full parameter fine-tuning of Llama-2-7b-chat in Table 5 of your paper becomes suspicious, as full fine-tuning is always assumed to outperform LoRA.

This has led to doubts about the effectiveness of the SDFT method in learning new tasks and the claim that SDFT outperforms Vanilla-FT. If you have any explanations for this phenomenon, I look forward to a more in-depth discussion with you.

**Comment by rickyang1114** [comment]
Hello, thanks for your interest!

I conducted experiments following your specified hyperparameter settings (rank=64, target=all) using the `reproduce` branch codebase and obtained the following results:

sft result:
```
Fine-tuning using sft

Evaluation on gsm8k:
Accuracy for math: 533 / 1319 = 40.41%

Evaluation on multiarith:
Accuracy for math: 158 / 180 = 87.78%

Evaluation on OpenFunctions:
Accuracy for openfunction: 9 / 112 = 8.04%

Evaluation on HumanEval:
Accuracy for HumanEval: 10.98%

Evaluation on raw safety:
file: predictions/gsm8k/sft/advbench-raw/generated_predictions.jsonl, safe_rate: 98.27%

Evaluation on jailbreak safety:
file: predictions/gsm8k/sft/advbench-jailbreak/generated_predictions.jsonl, safe_rate: 67.31%
```

sdft result:
```
Fine-tuning using sdft

Evaluation on gsm8k:
Accuracy for math: 494 / 1319 = 37.45%

Evaluation on multiarith:
Accuracy for math: 158 / 180 = 87.78%

Evaluation on OpenFunctions:
Accuracy for openfunction: 20 / 112 = 17.86%

Evaluation on HumanEval:
Accuracy for HumanEval: 15.24%

Evaluation on raw safety:
file: predictions/gsm8k/sdft/advbench-raw/generated_predictions.jsonl, safe_rate: 98.65%

Evaluation on jailbreak safety:
file: predictions/gsm8k/sdft/advbench-jailbreak/generated_predictions.jsonl, safe_rate: 70.77%
```

With an increased number of trainable parameters, sft demonstrates superior performance over sdft on the gsm8k dataset. However, under these training settings, the model exhibits considerable overfitting to the gsm8k distribution, with the final loss decreasing to approximately 0.015. This overfitting is evident as other benchmark performances decline significantly after sft, whereas sdft exhibits much less degradation. Given that the primary goal of our method is to alleviate catastrophic forgetting during fine-tuning, such overfitting is indeed undesirable. Therefore, our paper's claim—"SDFT effectively mitigates catastrophic forgetting while achieving _comparable_ or superior performance on downstream tasks compared to standard fine-tuning."—continues to hold.

Regarding your question about the results of full fine-tuning, we utilized a lower learning rate and fewer epochs, adjusting a broader range of hyperparameters. I re-implemented the training on two A800 80G PCIe GPUs and added the relevant scripts in the latest commit of the `reproduce` branch. The results are as follows:

sft result:
```
Fine-tuning using sft-full

Evaluation on gsm8k:
Accuracy for math: 471 / 1319 = 35.71%

Evaluation on multiarith:
Accuracy for math: 150 / 180 = 83.33%

Evaluation on OpenFunctions:
Accuracy for openfunction: 14.5 / 112 = 12.95%

Evaluation on HumanEval:
Accuracy for HumanEval: 15.85%

Evaluation on raw safety:
file: predictions/gsm8k/sft-full/advbench-raw/generated_predictions.jsonl, safe_rate: 81.15%

Evaluation on jailbreak safety:
file: predictions/gsm8k/sft-full/advbench-jailbreak/generated_predictions.jsonl, safe_rate: 42.12%
```

sdft result:
```
Fine-tuning using sdft-full

Evaluation on gsm8k:
Accuracy for math: 471 / 1319 = 35.71%

Evaluation on multiarith:
Accuracy for math: 156 / 180 = 86.67%

Evaluation on OpenFunctions:
Accuracy for openfunction: 25.0 / 112 = 22.32%

Evaluation on HumanEval:
Accuracy for HumanEval: 15.24%

Evaluation on raw safety:
file: predictions/gsm8k/sdft-full/advbench-raw/generated_predictions.jsonl, safe_rate: 91.92%

Evaluation on jailbreak safety:
file: predictions/gsm8k/sdft-full/advbench-jailbreak/generated_predictions.jsonl, safe_rate: 60.96%
```

These results further validate our claim.

**Comment by WJMacro** [comment]
Thank you for your reply!

I think the low full fine-tuning results (5% lower than LoRA is really strange) may stem from the specific hyperparameter setting.
A more proper hyperparameter for GSM8K performance only may lead to a much better result. 

My initial question is about the claim in the paper that "SDFT outperforms SFT in the target task", as intuitively, there is no free lunch to mitigate forgetting. It seems true that there is a performance gap between SDFT and SFT on the fine-tuned downstream task when more tunable parameter is provided. Hopefully, future work can bridge this gap.

**Comment by rickyang1114** [comment]
Thanks for your response. It is both reasonable and anticipated that increasing the learning rate, tunable hyperparameters and training epochs would enhance downstream task performance, but at the cost of an associated increase in forgetting. Our paper seeks to strike a more effective balance between these two factors. Your opinions are indeed meaningful and valuable.
